### DISCUSSION TITLE
Availability of `vtkImageResliceMapper` in `pvpython`

### DISCUSSION THREAD

**Discussion by Besler** [Kitware/trame#193] [Trame/ParaView] [answered]
I am trying to do remote rendering with [pvpython](https://github.com/Kitware/paraview-visualizer/tree/master/docker). My Dockerfile looks like the following:

```dockerfile
FROM kitware/trame:glvnd as base

# Install ParaView-EGL
ARG PV_URL=https://www.paraview.org/files/v5.11/ParaView-5.11.0-egl-MPI-Linux-Python3.9-x86_64.tar.gz
RUN mkdir -p /opt/paraview && cd /opt/paraview && wget -qO- $PV_URL | tar --strip-components=1 -xzv
ENV TRAME_PARAVIEW=/opt/paraview

# Copy the server directory
COPY --chown=trame-user:trame-user ./server /deploy/server
```

I get the following error:

```bash
    from vtkmodules.vtkRenderingImage import vtkImageResliceMapper
ModuleNotFoundError: No module named 'vtkmodules.vtkRenderingImage'
```
I know classes like `vtkImageResliceMapper` are not available in vtk.js. Is this the same with pvpython?

**Comment by jourdain** [comment, marked as answer]
That seems to be a VTK module that we do not enable for ParaView. I know we have VTK wheel in the work for OSMesa and EGL but don't know when they will be available on our website for download... 

In the meantime you may have to build your own EGL wheel of VTK.

**Reply by banesullivan** [reply]
So. Wheels! 

We've been working diligently to start building and publishing OSMesa wheels for VTK. These will be available sometime this week from the GitLab Package Registry for the VTK repo: https://gitlab.kitware.com/vtk/vtk/-/packages

We've named them `vtk_osmesa` to differentiate them from the default `vtk` wheel (or other build variants).

You can install with (when they get published later this week):

```bash
pip install vtk_osmesa --pre --no-cache --extra-index-url https://gitlab.kitware.com/api/v4/projects/13/packages/pypi/simple
```

## EGL

It sounds like you actually want an EGL build variant of VTK. We do not have "official" wheels for this, though I have streamlined the process for building these wheels (OSMesa and EGL) in this repo: https://github.com/banesullivan/vtk-cmake

The configuration files there make it super easy to get either OSMesa or EGL wheels for VTK that have module parity with what's available for the "default" VTK wheel on PyPI.

### Pre-built!

Additionally, I have built and published some wheels using the above configurations here: https://github.com/pyvista/pyvista-wheels

But be careful with that repo... there's some cruft in there from a while ago. Any wheels added in the last month or so should be good to use. However, they are not `manylinux` wheels (as I was lazy when building these)

You can install an EGL wheel for Python 3.11 on linux with:

```
pip install https://github.com/pyvista/pyvista-wheels/raw/main/vtk_egl-9.2.5-cp311-cp311-linux_x86_64.whl
```

### Building a wheel

I'm happy to do this maybe a couple of times... but I don't want to open the floodgates to people on the internet having me build them VTK wheels 😄... but if you are fine with a `linux_x86_64` wheel (not `manylinux`), I'd be happy to build a wheel for your desired Python version and add any additional modules you need.

Though! I'd like to encourage you to try the CMake configurations I shared to do this yourself and contribute back any modifications: https://github.com/banesullivan/vtk-cmake

For posterity, I'd like to also point you to [PyVista's documentation](https://dev.pyvista.org/extras/building_vtk.html) which has further instructions on building VTK wheels

**Reply by banesullivan** [reply]
And the OSMesa wheels are up:

- `master`: https://gitlab.kitware.com/vtk/vtk/-/packages/90
- `9.2.5`: https://gitlab.kitware.com/vtk/vtk/-/packages/92

**Reply by banesullivan** [reply]
To get the latest 9.2.x stable wheel:

```
pip install vtk_osmesa<=9.2.1000 --pre --no-cache --extra-index-url https://gitlab.kitware.com/api/v4/projects/13/packages/pypi/simple
```

**Reply by Besler** [reply]
Worked like magic. Here for poetry.
```bash
poetry source add --secondary kitware https://gitlab.kitware.com/api/v4/projects/13/packages/pypi/simple
poetry add --source kitware vtk_osmesa
```

**Comment by jourdain** [comment]
VTK egl wheel is out and available...

`pip install --extra-index-url https://wheels.vtk.org vtk-egl`
